source/common/upstream/deterministic_aperture_lb.h

```diff
@@ -0,0 +1,161 @@
+ *
+ * High level ideas:
+ *      - Conserve connections made to the backend/endpoints by dividing the backends uniformly
+ *      among participating peers. This way, we avoid a complete mesh of connections between Envoy
+ *      and backends.
+ * Implementation:
+ *      - Uses a `Hash Ring` for the endpoints.
+ *      - Envoy peers are placed on a ring that is conceptually laid on top of the Ring Hash of
+ * backends to determine their overlaps.
+ *      - By using the ring overlaps as a mechanism to divide the
+ * backends, the backends get divided as a fraction of their overlaps with each peer.
+ *      - The algorithm uses a uniform random distribution to select a backend from the range of
+ * overlapping backends. This along with P2C ensures uniform load distribution.
+ */
+
+/**
+ * All DeterministicAperture load balancer stats. @see stats_macros.h
+ */
+#define ALL_DETERMINISTIC_APERTURE_LOAD_BALANCER_STATS(COUNTER) COUNTER(pick2_errors)
+
+/**
+ * Struct definition for all DeterministicAperture load balancer stats. @see stats_macros.h
+ */
+struct DeterministicApertureLoadBalancerStats {
+  ALL_DETERMINISTIC_APERTURE_LOAD_BALANCER_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+/**
+ * Thread aware load balancer implementation for DeterministicAperture.
+ */
+class DeterministicApertureLoadBalancer : public RingHashLoadBalancer {
+public:
+  DeterministicApertureLoadBalancer(
+      const PrioritySet& priority_set, ClusterStats& stats, Stats::Scope& scope,
+      Runtime::Loader& runtime, Random::RandomGenerator& random,
+      const absl::optional<envoy::config::cluster::v3::Cluster::DeterministicApertureLbConfig>&
+          config,
+      const envoy::config::cluster::v3::Cluster::CommonLbConfig& common_config);
+
+  const DeterministicApertureLoadBalancerStats& stats() const { return stats_; }
+  const RingHashLoadBalancerStats& ringStats() const { return ring_stats_; }
+
+  static DeterministicApertureLoadBalancerStats generateStats(Stats::Scope& scope);
+
+  using HashFunction = envoy::config::cluster::v3::Cluster::RingHashLbConfig::HashFunction;
+
+  /*
+   * Customization of the `RingHashLoadBalancer` ring to add functionality that allows calculating
+   * the intersecting ring segments.
+   */
+  struct Ring : public RingHashLoadBalancer::Ring {
+    Ring(double offset, double width, const NormalizedHostWeightVector& normalized_host_weights,
+         double min_normalized_weight, uint64_t min_ring_size, uint64_t max_ring_size,
+         HashFunction hash_function, bool use_hostname_for_hashing,
+         RingHashLoadBalancerStats ring_hash_stats, DeterministicApertureLoadBalancerStats& stats);
+
+    // ThreadAwareLoadBalancerBase::HashingLoadBalancer
+    HostConstSharedPtr chooseHost(uint64_t hash, uint32_t attempt) const override;
+
+    //
+    // Utility Ring methods. Note that they are public so that we can add unit tests easily.
```

Comment:
I think we normally prefer not to expose methods just for testing. We have cases where we mark a test as a friend class for this kind of direct testing of class internals, but I wonder if it's not sufficient to provide a mocked RNG to test the different paths.